include/alp/omp/storagebasedmatrix.hpp

```diff
@@ -0,0 +1,172 @@
+			template<
+				typename MatrixType,
+				std::enable_if< internal::is_storage_based< MatrixType >::value > *
+			>
+			friend const typename MatrixType::amf_type &getAmf( const MatrixType &A ) noexcept;
+
+			public:
+
+				/** Expose static properties */
+
+				typedef T value_type;
+				typedef AmfType amf_type;
+				typedef typename AmfType::imf_r_type imf_r_type;
+				typedef typename AmfType::imf_c_type imf_c_type;
+				/** Type returned by access function */
+				typedef T &access_type;
+				typedef const T &const_access_type;
+				/** Type of the index used to access the physical storage */
+				typedef typename AmfType::storage_index_type storage_index_type;
+
+			protected:
+				typedef StorageBasedMatrix< T, AmfType, requires_allocation, omp > self_type;
+				friend MatrixBase< self_type >;
+
+				typedef typename std::conditional<
+					requires_allocation,
+					Vector< T, omp >,
+					Vector< T, omp > &
+				>::type container_type;
+
+				/** A container-type view is characterized by its association with a physical container */
+				container_type container;
+
+				/**
+				 * Access mapping function maps a pair of logical coordinates
+				 * into the concrete coordinate inside the actual container.
+				 * \see AMF
+				 */
+				AmfType amf;
+
+				/**
+				 * determines the size of the matrix via the domain of
+				 * the index mapping functions.
+				 *
+				 * @return A pair of dimensions.
+				 */
+				std::pair< size_t, size_t > dims() const noexcept {
+					return amf.getLogicalDimensions();
+				}
+
+				size_t getStorageDimensions() const noexcept {
+					return amf.getStorageDimensions();
+				}
+
+				friend const Vector< T, omp > &getContainer( const self_type &A ) {
+					return A.container;
+				}
+
+				friend Vector< T, omp > &getContainer( self_type &A ) {
+					return A.container;
+				}
+
+				bool getInitialized() const noexcept {
+					return internal::getInitialized( container );
+				}
+
+				void setInitialized( const bool initialized ) noexcept {
+					internal::setInitialized( container , initialized );
+				}
+
+				const AmfType &getAmf() const noexcept {
+					return amf;
+				}
+
+				/**
+				 * Returns a constant reference to the element corresponding to
+				 * the provided storage index.
+				 *
+				 * @param storageIndex  storage index in the physical iteration
+				 *                      space.
+				 *
+				 * @return const reference or value of the element at given position.
+				 */
+				const_access_type access( const storage_index_type &storageIndex ) const {
+					return getBuffer( container, storageIndex.buffer_id )[ storageIndex.offset ];
+				}
+
+				access_type access( const storage_index_type &storageIndex ) {
+					return getBuffer( container, storageIndex.buffer_id )[ storageIndex.offset ];
+				}
+
+				storage_index_type getStorageIndex( const size_t i, const size_t j, const size_t s, const size_t P ) const {
+					return amf.getStorageIndex( i, j, s, P );
```

Comment:
what are `s` and `P` here? I recall we where leaving them as placeholder for parallel concepts but now that we are implementing them I wonder if these became redundant.